eth2/simulation/src/simulation.rs

```diff
@@ -15,69 +18,441 @@ where
+                .execution_environment_states
+                .push(cloned_initial_state)
+                .map_err(|_| Error::MaxLengthExceeded {
+                    what: format!("number of execution environment states"),
+                })?;
+
+            // Each shard should have the same # of ee states as there are EEs
+            assert_eq!(
+                shard_state.execution_environment_states.len(),
+                self.store.current_beacon_state.execution_environments.len()
+            );
+        }
+
+        let ee_index = self.store.current_beacon_state.execution_environments.len() - 1;
+        Ok(ee_index as u64)
     }
 
     /// Add a new shard block containing a list of transactions that need to be executed
     /// Execute all transactions on the appropriate shards / EEs, return ShardBlock index
-    fn create_shard_block(args: args::CreateShardBlock) -> ShardSlot {
-        unimplemented!();
+    pub fn create_shard_block(&mut self, a: simulation_args::CreateShardBlock) -> Result<u64> {
+        // Get the specified ShardState (if it exists)
+        let shard_index = a.shard_index as usize;
+        let shard = Shard::new(a.shard_index);
+        let shard_state = self
+            .store
+            .current_beacon_state
+            .shard_states
+            .get_mut(shard_index)
+            .ok_or(Error::OutOfBounds {
+                what: WhatBound::Shard,
+                index: shard_index,
+            })?;
+
+        // Create the internal shard block from args
+        let shard_block: ShardBlock<T> = ShardBlock::try_from(a.shard_block).context(ArgsError)?;
+
+        // Execute transactions and update shard state for all transactions
+        for transaction in shard_block.transactions.iter() {
+            // Get the specified EE (if it exists)
+            let ee_index: usize = transaction.ee_index.into();
+            let execution_environment = self
+                .store
+                .current_beacon_state
+                .execution_environments
+                .get(ee_index)
+                .ok_or(Error::OutOfBounds {
+                    what: WhatBound::ExecutionEnvironment,
+                    index: ee_index,
+                })?;
+
+            // Get the current EE state
+            let pre_state = shard_state
+                .execution_environment_states
+                .get(ee_index)
+                .ok_or(Error::OutOfBounds {
+                    what: WhatBound::ExecutionEnvironmentState,
+                    index: ee_index,
+                })?;
+
+            // Create a new runtime with the EE code, transaction data, and pre state root
+            let wasm_code: &[u8] = &*execution_environment.wasm_code;
+            let data: &[u8] = &*transaction.data;
+            let pre_state: [u8; 32] = pre_state.clone().into();
+            let mut runtime = RootRuntime::new(wasm_code, data, pre_state);
+            let post_root = runtime.execute();
+            drop(runtime);
```

Comment:
I think you can avoid the explicit `drop`:

```rust
let post_root = RootRuntime::new(wasm_code, data, pre_state).execute();
```

Reply:
Good call, I'll update this.